L1Trigger/L1TZDC/plugins/L1TZDCProducer.cc

```diff
@@ -0,0 +1,4885 @@
+//         Created:  Thu, 05 Dec 2013 17:39:27 GMT
+//
+// Copied for ZDC by: Chris McGinn
+//        Copy Made: Wed, 03 Aug 2023
+//        Contact: christopher.mc.ginn@cern.ch or
+//                 cfmcginn on github for bugs/issues
+//
+#include <memory>
+#include <iostream>
+// user include files
+
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "L1Trigger/L1TCalorimeter/interface/Stage2Layer2FirmwareFactory.h"
+#include "L1Trigger/L1TCalorimeter/interface/Stage2MainProcessor.h"
+#include "L1Trigger/L1TCalorimeter/interface/CaloTools.h"
+
+#include "L1Trigger/L1TCalorimeter/interface/CaloParamsHelper.h"
+#include "CondFormats/DataRecord/interface/L1TCaloParamsRcd.h"
+#include "CondFormats/DataRecord/interface/L1TCaloParamsO2ORcd.h"
+
+#include "DataFormats/L1TCalorimeter/interface/CaloTower.h"
+#include "DataFormats/L1Trigger/interface/EGamma.h"
+#include "DataFormats/L1Trigger/interface/Tau.h"
+#include "DataFormats/L1Trigger/interface/Jet.h"
+#include "DataFormats/L1Trigger/interface/EtSum.h"
+
+#include "DataFormats/HcalDigi/interface/HcalDigiCollections.h"
+
+//
+// class declaration
+//
+
+using namespace l1t;
+
+class L1TZDCProducer : public edm::stream::EDProducer<> {
+public:
+  explicit L1TZDCProducer(const edm::ParameterSet& ps);
+  ~L1TZDCProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  void endRun(edm::Run const&, edm::EventSetup const&) override;
+
+  int zdcLUTIndexHelper(int iDetPos, int iBXPos);
+
+  //virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+  //virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+
+  // ----------member data ---------------------------
+
+  // input tokens
+  //Add the ZDC token - remove other tokens later when more certain what is not needed i.e. at direction of L1Trigger maintainers, particularly the Stage2Layer2Producer maintainers
+  edm::EDGetTokenT<QIE10DigiCollection> m_zdcToken;
+  edm::ESGetToken<CaloParams, L1TCaloParamsRcd> m_candidateToken;
+
+  // put tokens
+  edm::EDPutTokenT<EtSumBxCollection> m_etToken;
+
+  //Following the L1TStage2Layer2Producer
+  CaloParamsHelper* m_params;
```

Comment:
Please use a smart pointer instead of explicit `new`/`delete`:
```suggestion
  std::unique_ptr<CaloParamsHelper> m_params;
```

If `CaloParamsHelper` is small enough (a few hundred bytes ?) you can just use it by value:
```suggestion
  CaloParamsHelper m_params;
```

If `CaloParamsHelper` is large and the same identical object is used by multiple `EDProducer`s, it would be better to move it to an `ESProduct`.

Reply:
I have opted to use the smart pointer solution; tests appear to work with this implementation. This is in the latest commit